### PR TITLE
Stream chain updates through the chain index.

### PR DESCRIPTION
**Pull request by raduom**
To make the ChainIndex sync with a alonzo node we had to:
* Make all messages involving traces serialisable to JSON and providing ToObject instances.
* Add support for Byron Addresses (so we can track outputs for Byron transactions)

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

### CODE REVIEW

File: plutus-chain-index/app/Main.hs [hunk 1]
```diff
@@ -9,19 +9,101 @@
...
+    let ciBlock = fromCardanoBlock block
+    case ciBlock of
+      Left err    ->
+        logError trace (ConversionFailed err)
+      Right block' ->
+        runChainIndex trace mState $ appendBlock (fromCardanoTip tip) block'
+chainSyncHandler trace mState
+  (RollBackward _ tip) _ =
+    -- Do we really want to pass the tip of the new blockchain to the
+    -- rollback function (rather than the point where the chains diverge)?
```
Comment (j-mueller) [inline plutus-chain-index/app/Main.hs:103]: I thought `tip` _was_ the point where the chains diverge?

File: plutus-chain-index/src/Plutus/Contract/CardanoAPI.hs [hunk 2]
```diff
@@ -184,7 +192,16 @@ toCardanoTxOut networkId (P.TxOut addr value datumHash) =
             <*> toCardanoTxOutDatumHash datumHash
 
 fromCardanoAddress :: C.AddressInEra era -> Either FromCardanoError P.Address
-fromCardanoAddress (C.AddressInEra C.ByronAddressInAnyEra _) = Left ByronAddressesNotSupported
+fromCardanoAddress (C.AddressInEra C.ByronAddressInAnyEra (C.ByronAddress address)) =
```
Comment (j-mueller) [inline plutus-chain-index/src/Plutus/Contract/CardanoAPI.hs:195]: Don't Byron addresses have a different length? I'm worried that this stops working when we convert the `P.Address` back to a `cardano-api` address. We can leave it in for now, but I'll try use the original `C.AddressInEra` in more places.

**Review by j-mueller — APPROVED** [review]

File: plutus-chain-index/app/Main.hs (hunk 1 above)
Comment (raduom) [reply plutus-chain-index/app/Main.hs:103]: There's a bunch of errors that I guess stem from this. I will fix it in the let's test things PR.

File: plutus-chain-index/src/Plutus/Contract/CardanoAPI.hs (hunk 2 above)
Comment (raduom) [reply plutus-chain-index/src/Plutus/Contract/CardanoAPI.hs:195]: So we need something in `P.Address` to say which era the address comes from? I think there would be a bunch of changes to be made, so I think a new PR would be welcomed for it. Feel free to assign me.

Comment (j-mueller) [reply plutus-chain-index/src/Plutus/Contract/CardanoAPI.hs:195]: Yeah probably. We basically need an exact copy of their address type

**Pull request merged**